### PR TITLE
Confirmation Dialog refactor and line input

### PR DESCRIPTION
**Pull request by shermp**
In order to support text input, I decided to refactor the Confirmation Dialog code into its own class.

I've then added support for adding a line input dialog to get input from the user. The line input can be normal or a password field, complete with *Show Password* checkbox. It is also possible to add placeholder text when creating the input box.

Two new methods `dlgConfirmLineEdit` and `dlgConfirmLineEditPlaceholder` have been created. A new signal, `dlgConfirmTextInput` has also been created.

When the user taps the *accept* button, the `dlgConfirmTextInput` signal is emitted with the contents of the input box, otherwise `dlgConfirmResult` is emitted.

During my experiments, I had a multi-line text edit dialog working, but decided to drop it in favor of the current implementation which is far simpler and hopefully a lot less fragile, but only supports single line input.

See also #9 for my experiment musings and results.

Tested on FW 4.6 and 4.25. One hopes it will work on the firmware versions between those tested...

### CODE REVIEW

File: src/ndb/NDBCfmDlg.cc [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    /* Resolve symbols */
+    // Confirmation Dialog
+    NDB_RESOLVE_SYMBOL("_ZN25ConfirmationDialogFactory21getConfirmationDialogEP7QWidget", nh_symoutptr(symbols.ConfirmationDialogFactory_getConfirmationDialog));
+    NDB_RESOLVE_SYMBOL("_ZN25ConfirmationDialogFactory18showTextEditDialogERK7QString", nh_symoutptr(symbols.ConfirmationDialogFactory_showTextEditDialog));
+    NDB_RESOLVE_SYMBOL("_ZN18ConfirmationDialog8setTitleERK7QString", nh_symoutptr(symbols.ConfirmationDialog__setTitle));
+    NDB_RESOLVE_SYMBOL("_ZN18ConfirmationDialog7setTextERK7QString", nh_symoutptr(symbols.ConfirmationDialog__setText));
+    NDB_RESOLVE_SYMBOL("_ZN18ConfirmationDialog19setAcceptButtonTextERK7QString", nh_symoutptr(symbols.ConfirmationDialog__setAcceptButtonText));
+    NDB_RESOLVE_SYMBOL("_ZN18ConfirmationDialog19setRejectButtonTextERK7QString", nh_symoutptr(symbols.ConfirmationDialog__setRejectButtonText));
+    NDB_RESOLVE_SYMBOL("_ZN18ConfirmationDialog15showCloseButtonEb", nh_symoutptr(symbols.ConfirmationDialog__showCloseButton));
+    NDB_RESOLVE_SYMBOL("_ZN18ConfirmationDialog21setRejectOnOutsideTapEb",nh_symoutptr(symbols.ConfirmationDialog__setRejectOnOutsideTap));
```
Comment (pgaskin) [inline src/ndb/NDBCfmDlg.cc:32]: Missing space.

File: src/ndb/NDBCfmDlg.cc [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+NDBCfmDlg::NDBCfmDlg(QObject* parent, void* libnickel) : QObject(parent) {
+    initResult = Ok;
+    if (!libnickel) {
+        initResult = InitError;
+        return;
+    }
+    currActiveType = TypeStd;
+    /* Resolve symbols */
```
Comment (pgaskin) [inline src/ndb/NDBCfmDlg.cc:23]: You might want to consider adding the NM symtest comments and running it in GH Actions now that you use quite a few symbols in NDB.

File: src/ndb/NDBCfmDlg.cc [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        ret = tle->text();
+    }
+    return ret;
+}
+
+void NDBCfmDlg::setText(QString const& text) {
+    if (dlg && currActiveType == TypeLineEdit && tle) {
+        tle->setPlaceholderText(text);
+    }
+}
```
Comment (pgaskin) [inline src/ndb/NDBCfmDlg.cc]: Line ending.

File: src/ndb/NDBCfmDlg.cc [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+            else {
+                tcb->setChecked(true);
+                tcb->hide();
+            }
+        }
+    }
+}
+
+QString NDBCfmDlg::getText() {
+    QString ret("");
```
Comment (pgaskin) [inline src/ndb/NDBCfmDlg.cc:157]: You don't need the initializer there.

File: src/ndb/ndb.cc [hunk 5]
```diff
@@ -437,7 +414,10 @@ void NDB::dlgConfirmation(QString const& title, QString const& body, QString con
  * When the dialog is closed, a \l dlgConfirmResult() signal is emitted.
  */
 void NDB::dlgConfirmNoBtn(QString const& title, QString const& body) {
-    return dlgConfirmation(title, body, QString(""), QString(""), true, true);
+    NDB_DBUS_USB_ASSERT((void) 0);
+    NDB_DLG_ASSERT((void) 0, (cfmDlg->createDialog(NDBCfmDlg::TypeStd, title, body, QString(""), QString(""), true) == NDBCfmDlg::Ok));
```
Comment (pgaskin) [inline src/ndb/ndb.cc:418]: You don't need the initializer for those `QString`s either.

File: src/ndb/ndb.cc [hunk 6]
```diff
@@ -450,7 +430,10 @@ void NDB::dlgConfirmNoBtn(QString const& title, QString const& body) {
  * \l dlgConfirmResult() signal is emitted.
  */
 void NDB::dlgConfirmAccept(QString const& title, QString const& body, QString const& acceptText) {
-    return dlgConfirmation(title, body, acceptText, QString(""), true, true);
+    NDB_DBUS_USB_ASSERT((void) 0);
+    NDB_DLG_ASSERT((void) 0, (cfmDlg->createDialog(NDBCfmDlg::TypeStd, title, body, acceptText, QString(""), true) == NDBCfmDlg::Ok));
```
Comment (pgaskin) [inline src/ndb/ndb.cc:434]: Same here.

File: src/ndb/ndb.cc [hunk 7]
```diff
@@ -463,7 +446,10 @@ void NDB::dlgConfirmAccept(QString const& title, QString const& body, QString co
  * \l dlgConfirmResult() signal is emitted.
  */
 void NDB::dlgConfirmReject(QString const& title, QString const& body, QString const& rejectText) {
-    return dlgConfirmation(title, body, QString(""), rejectText, true, true);
+    NDB_DBUS_USB_ASSERT((void) 0);
+    NDB_DLG_ASSERT((void) 0, (cfmDlg->createDialog(NDBCfmDlg::TypeStd, title, body, QString(""), rejectText, true) == NDBCfmDlg::Ok));
```
Comment (pgaskin) [inline src/ndb/ndb.cc:450]: And here. All of the other ones too.

**Review by pgaskin — COMMENTED** [review]
Haven't looked at the logic yet.

File: src/ndb/NDBCfmDlg.cc (hunk 3 above)
Comment (shermp) [reply src/ndb/NDBCfmDlg.cc]: Already done. I caught it after reviewing my own PR

File: src/ndb/NDBCfmDlg.cc (hunk 2 above)
Comment (shermp) [reply src/ndb/NDBCfmDlg.cc:23]: Not a bad idea, but I'll probably do it outside this PR.

File: src/ndb/ndb.cc (hunk 5 above)
Comment (shermp) [reply src/ndb/ndb.cc]: Done

File: src/ndb/ndb.cc (hunk 7 above)
Comment (shermp) [reply src/ndb/ndb.cc]: And done

File: src/ndb/ndb.cc (hunk 6 above)
Comment (shermp) [reply src/ndb/ndb.cc]: Done

File: src/ndb/NDBCfmDlg.cc (hunk 4 above)
Comment (shermp) [reply src/ndb/NDBCfmDlg.cc]: Done

File: src/ndb/NDBCfmDlg.cc (hunk 1 above)
Comment (shermp) [reply src/ndb/NDBCfmDlg.cc]: Fixed

File: src/ndb/ndb.cc (hunk 7 above)
Comment (pgaskin) [reply src/ndb/ndb.cc]: I was actually referring to the `""`, as in to change them to `QString()`, but what you did is fine too (it's technically slightly different, but it won't make a difference in practice).

Comment (shermp) [reply src/ndb/ndb.cc]: Yeah, thought about it, looked it up, saw that using a bare literal was ok, so went with it.

**Comment by shermp** [comment]
As another note, I never did end up figuring out how to hook up the keyboard's `Go` button, so tapping that does nothing.

**Comment by shermp** [comment]
> As another note, I never did end up figuring out how to hook up the keyboard's `Go` button, so tapping that does nothing.

Well, that was easy. Turns out that `N3ConfirmationTextEditField::commitRequested()` is the signal that is emitted when the `Go` button is tapped. 

Let me implement that properly. Just need connect it to the `accepted` signal of the dialog and should be good to go.

**Comment by shermp** [comment]
Apologies for the late changes. Think I'm done now.

**Comment by shermp** [comment]
Any chance of testing/reviewing this in the near future @pgaskin and @NiLuJe ?

**Comment by pgaskin** [comment]
I want to look at this closely, but I might not be able to get around to it until Sunday (I'll be making the NM release around the same time).

**Comment by shermp** [comment]
That's fine, thanks for the update.

I'm quite happy for you to have a close look. Don't want to be releasing something with any major issues after all.

**Comment by shermp** [comment]
Closing this for now in favor of #11

**Pull request closed without merging**